src/lua/compat.lua

```diff
@@ -0,0 +1,333 @@
+Available commands:
+
+    compat.help()                   -- show this help
+    compat.<option_name>            -- list option info
+    compat.<option_name> = 'old'    -- set desired value to option, could be 'old', 'new', 'default'
+    compat{<option_name> = 'new'}   -- set listed options to desired values ('old', 'new', 'default')
+    compat.dump()                   -- get Lua command that sets up different compat with same options as current
+    compat.add_option()             -- add new option by providing a table with name, default, brief, obsolete
+                                       and action function
+]]
+
+-- Returns table with all non-obsolete options from `options` and their values.
+local function serialize_compat()
+    -- The results of serialization should be ordered correctly. The only
+    -- feasible way for it is using indexed table with values {option_name = val}.
+    local result = { }
+
+    for _, name in pairs(options_order) do
+        local option = options[name]
+        if option.selected and option.current == NEW and not option.obsolete then
+            table.insert(result, {[name] = 'new'})
+        end
+    end
+    for _, name in pairs(options_order) do
+        local option = options[name]
+        if option.selected and option.current == OLD and not option.obsolete then
+            table.insert(result, {[name] = 'old'})
+        end
+    end
+    for _, name in pairs(options_order) do
+        local option = options[name]
+        if not option.selected and not option.obsolete then
+            table.insert(result, {[name] = ('default (%s)'):format(option.current and 'new' or 'old')})
+        end
+    end
+
+    -- Result is an indexed table but this way user can interact with it as if
+    -- it was key-value.
+    result = setmetatable(result, {
+        __index = function(self, key)
+            for i = 1, #self do
+                if self[i][key] ~= nil then
+                    return self[i][key]
+                end
+            end
+        end
+    })
+
+    return result
+end
+
+-- Checks options correctness.
+local function verify_option(name, option)
+    if type(name) ~= 'string' then
+        error(('Option name must be a string (%s provided)'):format(type(name)))
+    end
+    for p, t in pairs(options_format) do
+        if not t:find(type(option[p])) then
+            error(('Invalid option table for %s, bad %s (%s is expected)'):format(name, p, t))
+        end
+    end
+    if option.default ~= 'new' and option.default ~= 'old' then
+        error(("Invalid option table for %s, bad default ('new'/'old' is expected)"):format(name))
+    end
+    if option.obsolete and option.default == 'old' then
+        error(('Obsolete option %s default is wrong.'):format(name))
+    end
+    if type(option.run_action) ~= 'boolean' and option.run_action ~= nil then
+        error(("Invalid option table for %s, bad run_action (true/false is expected)"):format(name))
+    end
```

Comment:
Isn't it verified in the `if` in the loop above?

Reply:
Fixed.